kani-compiler/src/codegen_cprover_gotoc/codegen/statement.rs

```diff
@@ -467,32 +458,90 @@ impl<'tcx> GotocCtx<'tcx> {
+    /// details.
+    fn extract_ptr(&self, arg_expr: Expr, arg_ty: Ty<'tcx>) -> Expr {
+        // Generate an expression that indexes the pointer.
+        let expr = self
+            .receiver_data_path(arg_ty)
+            .fold(arg_expr, |curr_expr, (name, _)| curr_expr.member(name, &self.symbol_table));
+
+        trace!(?arg_ty, gotoc_ty=?expr.typ(), gotoc_expr=?expr.value(), "extract_ptr");
+        expr
+    }
+
+    /// Codegen the dynamic call to a trait method via the fat pointer vtable.
+    ///
+    /// If the original call was of the form
+    /// f(arg0, arg1);
+    ///
+    /// The new call should be of the form
+    /// arg0.vtable->f(arg0.data,arg1);
+    ///
+    /// For that, we do the following:
+    /// 1. Extract the fat pointer out of the first argument.
+    /// 2. Obtain the function pointer out of the fat pointer vtable.
+    /// 3. Change the first argument to only reference the data pointer (instead of the fat one).
+    ///     - When the receiver type is a `struct` we need to build a structure that mirrors
+    ///       the original one but uses a thin pointer instead.
+    /// 4. Generate the function call.
     fn codegen_virtual_funcall(
         &mut self,
-        trait_fat_ptr: Expr,
+        self_ty: Ty<'tcx>,
         def_id: DefId,
         idx: usize,
         place: &Place<'tcx>,
         fargs: &mut [Expr],
         loc: Location,
     ) -> Vec<Stmt> {
         let vtable_field_name = self.vtable_field_name(def_id, idx);
+        trace!(?self_ty, ?place, ?vtable_field_name, "codegen_virtual_funcall");
+
+        let trait_fat_ptr = self.extract_ptr(fargs[0].clone(), self_ty);
+        assert!(
+            trait_fat_ptr.typ().is_rust_trait_fat_ptr(&self.symbol_table),
+            "Expected fat pointer, but got {:?}",
+            trait_fat_ptr.typ()
+        );
 
-        // Now that we have all the stuff we need, we can actually build the dynamic call
-        // If the original call was of the form
-        // f(arg0, arg1);
-        // The new call should be of the form
-        // arg0.vtable->f(arg0.data,arg1);
         let vtable_ref = trait_fat_ptr.to_owned().member("vtable", &self.symbol_table);
         let vtable = vtable_ref.dereference();
         let fn_ptr = vtable.member(vtable_field_name, &self.symbol_table);
-
-        // Update the argument from arg0 to arg0.data
-        fargs[0] = trait_fat_ptr.to_owned().member("data", &self.symbol_table);
+        trace!(fn_typ=?fn_ptr.typ(), "codegen_virtual_funcall");
+
+        let data_ptr = trait_fat_ptr.to_owned().member("data", &self.symbol_table);
+        let mut ret_stmts = Vec::<Stmt>::new();
+        fargs[0] = if self_ty.is_adt() {
+            // Generate a temp variable and assign its inner pointer to the fat_ptr.data.
+            match fn_ptr.typ() {
+                Type::Pointer { typ } => match typ.as_ref() {
+                    Type::Code { parameters, .. } => {
```

Comment:
Is it possible to nest these two?

Reply:
I think the only way to do that is by using `#![feature(box_syntax)]` because `typ` is a `Box`.

Reply:
I guess we already have that feature enabled. I'll make the change you suggested. Thanks